services/motion/explore/explore.go

```diff
@@ -0,0 +1,571 @@
+// Package explore implements a motion service for exploration. This motion service model is a temporary model
+// that will be incorporated into the builtIn service in the future.
+package explore
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/edaniels/golog"
+	geo "github.com/kellydunn/golang-geo"
+	"github.com/pkg/errors"
+	servicepb "go.viam.com/api/service/motion/v1"
+	goutils "go.viam.com/utils"
+
+	"go.viam.com/rdk/components/base"
+	"go.viam.com/rdk/components/base/kinematicbase"
+	"go.viam.com/rdk/components/camera"
+	"go.viam.com/rdk/internal"
+	"go.viam.com/rdk/motionplan"
+	"go.viam.com/rdk/operation"
+	"go.viam.com/rdk/referenceframe"
+	"go.viam.com/rdk/resource"
+	"go.viam.com/rdk/robot/framesystem"
+	"go.viam.com/rdk/services/motion"
+	"go.viam.com/rdk/services/vision"
+	"go.viam.com/rdk/spatialmath"
+	"go.viam.com/rdk/utils"
+)
+
+var (
+	model            = resource.DefaultModelFamily.WithModel("explore")
+	errUnimplemented = errors.New("unimplemented")
+	// Places a limit on how far a potential move action can be performed.
+	moveLimit = 10000.
+	// The distance a detected obstacle can be from a base to trigger the Move command to stop.
+	validObstacleDistanceMM = 1000.
+)
+
+func init() {
+	resource.RegisterDefaultService(
+		motion.API, model,
+		resource.Registration[motion.Service, *Config]{
+			Constructor: NewExplore,
+			WeakDependencies: []internal.ResourceMatcher{
+				internal.ComponentDependencyWildcardMatcher,
+				internal.VisionDependencyWildcardMatcher,
+			},
+		})
+}
+
+const (
+	exploreOpLabel = "explore-motion-service"
+)
+
+// moveResponse is the message type returned by both channels: obstacle detection and plan execution.
+// It contains a bool stating if an obstacle was found or the execution was complete and/or any
+// associated error.
+type moveResponse struct {
+	err     error
+	success bool
+}
+
+// inputEnabledActuator is an actuator that interacts with the frame system.
+// This allows us to figure out where the actuator currently is and then
+// move it. Input units are always in meters or radians.
+type inputEnabledActuator interface {
+	resource.Actuator
+	referenceframe.InputEnabled
+}
+
+// obstacleDetectorObject provides a map for matching vision services to any and all cameras to be used with them.
+type obstacleDetectorObject map[vision.Service]camera.Camera
```

Comment:
[q] why is it `camera.Camera`? If we are only using the `GetObjectPointClouds` method this can just be `resource.Name`.

[q/nit]: given the description of this type wouldn't it be better to have this be `map[vision.Service][]resource.Name`?

Reply:
Good idea, do we need to cast to camera.Camera if we are just storing the name (casting will allow us to confirm that the camera component exists). See function `createObstacleDetectors`